hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/DataSkippingUtils.scala

```diff
@@ -29,148 +30,186 @@ import org.apache.spark.sql.functions.col
       case Not(EqualTo(value: Literal, attribute: AttributeReference)) =>
-        val colName = getTargetColNameParts(attribute)
-        reWriteCondition(colName, Not(And(EqualTo(minValue(colName), value), EqualTo(maxValue(colName), value))))
-      // query filter "colA not like xxxx"   convert it to "not ( colA_minValue startWith xxx and colA_maxValue startWith xxx)" for index table
+        val colName = getTargetColName(attribute, indexSchema)
+        Not(colContainsValuesEqualToLiteral(colName, value))
+      // Filter "colA not like xxx"
+      // Translates to "!(colA_minValue <= xxx AND colA_maxValue >= xxx)" for index lookup
+      // NOTE: This is a inversion of "colA like xxx" assuming that colA is a string-based type
       case Not(StartsWith(attribute, value @ Literal(_: UTF8String, _))) =>
-        val colName = getTargetColNameParts(attribute)
-        reWriteCondition(colName, Not(And(StartsWith(minValue(colName), value), StartsWith(maxValue(colName), value))))
+        val colName = getTargetColName(attribute, indexSchema)
+        Not(colContainsValuesEqualToLiteral(colName, value))
+
       case or: Or =>
-        val resLeft = createZindexFilter(or.left, indexSchema)
-        val resRight = createZindexFilter(or.right, indexSchema)
+        val resLeft = createZIndexLookupFilter(or.left, indexSchema)
+        val resRight = createZIndexLookupFilter(or.right, indexSchema)
         Or(resLeft, resRight)
 
       case and: And =>
-        val resLeft = createZindexFilter(and.left, indexSchema)
-        val resRight = createZindexFilter(and.right, indexSchema)
+        val resLeft = createZIndexLookupFilter(and.left, indexSchema)
+        val resRight = createZIndexLookupFilter(and.right, indexSchema)
         And(resLeft, resRight)
 
       case expr: Expression =>
         Literal.TrueLiteral
     }
   }
 
-  /**
-    * Extracts name from a resolved expression referring to a nested or non-nested column.
-    */
-  def getTargetColNameParts(resolvedTargetCol: Expression): Seq[String] = {
+  private def checkColIsIndexed(colName: String, indexSchema: StructType): Boolean = {
+    Set.apply(
+      getMinColumnNameFor(colName),
+      getMaxColumnNameFor(colName),
+      getNumNullsColumnNameFor(colName)
+    )
+      .forall(stat => indexSchema.exists(_.name == stat))
+  }
+
+  private def getTargetColName(resolvedExpr: Expression, indexSchema: StructType): String = {
+    val colName = UnresolvedAttribute(getTargetColNameParts(resolvedExpr)).name
+
+    // Verify that the column is indexed
+    // NOTE: That in case filtering expression contains referenced to a column that is NOT
+    //       indexed data pruning becomes (largely) impossible
+    if (!checkColIsIndexed(colName, indexSchema)) {
+      logDebug(s"Filtering expression contains column that is not indexed ($colName)")
+      throw new AnalysisException(s"Filtering expression contains column that is not indexed ($colName)")
```

Comment:
This place should not throw exceptions. For a filter, if the column it refers to does not exist in the index, just change the filter to literal.trueliteral.  View the original logic of rewritecondition， In this way, data will not be lost in the process of index query

Reply:
Yeah, this change is not by accident: when re-writing condition as `TrueLiteral` it actually returns incorrect results -- let's consider following scenario:

We have a Table T with Z-ordered columns A, B

Now, if we query `SELECT * FROM T WHERE A = ... OR B = ... OR C = ...` this will return incorrect results b/c C is not indexed and we'd data-pruning based on A and B only.

We can't do data-pruning if in the query there are present columns that aren't indexed*. 

*We actually can in some cases but for simplicity of making sure that we do things correctly i'm currently conservatively removing any of such opts

Reply:
Sorry, I may not agree with you here.  select * from T where A=  or B =  C=  ,  those fitler will be rewrite to  A = xx or b= xx or literal.trueliteral.   Due to the existence of this  literal.trueliteral we can push those fitler to index table directly,  This will not cause any problems。

Reply:
It will cause problems: in essence, we are relying on index to tell us the _trimmed set_ of files that we need to scan, right? Now if query does contain filter on the column that is **not indexed**, we can't use index any more b/c in the current flow we'd rely on set of files provided by the index as the only files we will be scanning, therefore missing on the rows that match criteria `C = ...`

Reply:
sorrry， i still cannot understand，  why we will missing on the rows that match criteria C = ...  could you give me a example.

Reply:
Surely: let's imagine that we have three files F1, F2, F3 all of which are _indexed_ on clustered columns A and B.

Now, imagine that i'm executing query that would do `WHERE A = ... OR B = ... OR C = ...`, if F3 does not satisfy `A = ... OR B ...` condition, we will skip right through it, though it could contain columns satisfying `C = ...` condition.

Reply:
no， WHERE A = ... OR B = ... OR C = ...  will be convert to   A condition Or B condiation Or  Literal.TrueLiteral
since this is an Or condition；  Literal.TrueLiteral will suppress other filter condition。  F3 will also include， no files will missing.

Reply:
My understanding/expectation is also leaning more towards what @xiarixiaoyao mentions. 
@alexeykudinkin are you saying that's not how we convert the query predicate to the query on the index table?

Reply:
@xiarixiaoyao i see what you were saying now.

Reply:
Addressed